akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpEntity.scala

```diff
@@ -354,4 +389,51 @@ object HttpEntity {
     def getTrailerHeaders: JIterable[jm.HttpHeader] = trailer.asJava
   }
   object LastChunk extends LastChunk("", Nil)
+
+  /**
+   * Turns the given source into one that respects the `withSizeLimit` calls when used as a parameter
+   * to entity constructors.
+   */
+  def limitable[Out: SizeOfInBytes, Mat](source: Source[Out, Mat]): Source[Out, Mat] =
+    source.transform { () ⇒
+      new PushStage[Out, Out] {
+        val sizeOf = implicitly[SizeOfInBytes[Out]]
+        var maxBytes = -1L
+        var bytesLeft = Long.MaxValue
+
+        override def preStart(ctx: LifecycleContext) =
+          ctx.attributes.getFirst[SizeLimit] match {
+            case Some(SizeLimit(bytes, cl @ Some(contentLength))) ⇒
+              if (contentLength > bytes) throw EntityStreamSizeException(bytes, cl)
+            // else we still count but never throw an error
+            case Some(SizeLimit(bytes, None)) ⇒
+              maxBytes = bytes
+              bytesLeft = bytes
+            case None ⇒
+          }
+
+        def onPush(elem: Out, ctx: stage.Context[Out]): stage.SyncDirective = {
+          bytesLeft -= sizeOf(elem)
+          if (bytesLeft >= 0) ctx.push(elem)
+          else ctx.fail(EntityStreamSizeException(maxBytes))
+        }
+      }
+    }
```

Comment:
What is wrong with two nested ´via`s? i don't get it. What you included as a snippet is the proper way to name a stage. Since every immediate step of a Flow is reusable, you need the .via() on the source, otherwise the name would appliy to the whole assembled Source. The first via (from transform) will not have any more overhead than creating a Flow direcly, since you start from a cached empty flow which will simply delegate to a new Flow. The second via, attaching to the Source has exactly the same cost as calling `.map` since that is implemented in terms of via as well.